cms/lib/xblock/upstream_sync.py

```diff
@@ -0,0 +1,442 @@
+            continue
+
+        # Fetch the upstream's value and save it on the downstream (ie, `downstream.upstream_$FIELD`).
+        old_upstream_value = getattr(downstream, fetch_field_name)
+        new_upstream_value = getattr(upstream, field_name)
+        setattr(downstream, fetch_field_name, new_upstream_value)
+
+        if only_fetch:
+            continue
+
+        # Okay, now for the nuanced part...
+        # We need to update the downstream field *iff it has not been customized**.
+        # Determining whether a field has been customized will differ in Beta vs Future release.
+        # (See "PRESERVING DOWNSTREAM CUSTOMIZATIONS" comment below for details.)
+
+        ## FUTURE BEHAVIOR: field is "customized" iff we have noticed that the user edited it.
+        #  if field_name in downstream.downstream_customized:
+        #      continue
+
+        ## BETA BEHAVIOR: field is "customized" iff we have the prev upstream value, but field doesn't match it.
+        downstream_value = getattr(downstream, field_name)
+        if old_upstream_value and downstream_value != old_upstream_value:
+            continue  # Field has been customized. Don't touch it. Move on.
+
+        # Field isn't customized -- update it!
+        setattr(downstream, field_name, new_upstream_value)
+
+
+def _update_non_customizable_fields(*, upstream: XBlock, downstream: XBlock) -> None:
+    """
+    For each field `downstream.blah` that isn't customizable: set it to `upstream.blah`.
+    """
+    syncable_fields = _get_synchronizable_fields(upstream, downstream)
+    customizable_fields = set(downstream.get_customizable_fields().keys())
+    for field_name in syncable_fields - customizable_fields:
+        new_upstream_value = getattr(upstream, field_name)
+        setattr(downstream, field_name, new_upstream_value)
+
+
+def _get_synchronizable_fields(upstream: XBlock, downstream: XBlock) -> set[str]:
+    """
+    The syncable fields are the ones which are content- or settings-scoped AND are defined on both (up,down)stream.
+    """
+    return set.intersection(*[
+        set(
+            field_name
+            for (field_name, field) in block.__class__.fields.items()
+            if field.scope in [Scope.settings, Scope.content]
+        )
+        for block in [upstream, downstream]
+    ])
+
+
+def decline_sync(downstream: XBlock) -> None:
+    """
+    Given an XBlock that is linked to upstream content, mark the latest available update as 'declined' so that its
+    authors are not prompted (until another upstream version becomes available).
+
+    Does not save `downstream` to the store. That is left up to the caller.
+
+    If `downstream` lacks a valid+supported upstream link, this raises an UpstreamLinkException.
+    """
+    # Try to load the upstream.
+    upstream_link = UpstreamLink.get_for_block(downstream)  # Can raise UpstreamLinkException
+    if not upstream_link:
+        return  # No upstream -> nothing to sync.
```

Comment:
```suggestion
    upstream_link = UpstreamLink.get_for_block(downstream)  # Can raise UpstreamLinkException
```

Forgot to simplify this block, now that `get_for_block` is guaranteed to return non-None.

Reply:
Done